File: packages/python-sdk/usesend/usesend.py
"""Core client for interacting with the UseSend API.

Enhancements:
- Optional ``raise_on_error`` to raise ``UseSendHTTPError`` on non-2xx.
- Reusable ``requests.Session`` support for connection reuse.
"""
from __future__ import annotations

import os
from typing import Any, Dict, Optional, Tuple

import requests


DEFAULT_BASE_URL = "https://app.usesend.com"


class UseSendHTTPError(Exception):
    """HTTP error raised when ``raise_on_error=True`` and a request fails."""

    def __init__(self, status_code: int, error: Dict[str, Any], method: str, path: str) -> None:
        self.status_code = status_code
        self.error = error
        self.method = method
        self.path = path
        super().__init__(self.__str__())

    def __str__(self) -> str:  # pragma: no cover - presentation only
        code = self.error.get("code", "UNKNOWN_ERROR")
        message = self.error.get("message", "")
        return f"{self.method} {self.path} -> {self.status_code} {code}: {message}"


class UseSend:
    """UseSend API client.

    Parameters
    ----------
    key:
        API key issued by UseSend. If not provided, the client attempts to
        read ``USESEND_API_KEY`` or ``UNSEND_API_KEY`` from the environment.
    url:
        Optional base URL for the API (useful for testing).
    """

    def __init__(
        self,
        key: Optional[str] = None,
        url: Optional[str] = None,
        *,
        raise_on_error: bool = True,
        session: Optional[requests.Session] = None,
    ) -> None:
        self.key = key or os.getenv("USESEND_API_KEY") or os.getenv("UNSEND_API_KEY")
        if not self.key:
            raise ValueError("Missing API key. Pass it to UseSend('us_123')")

        base = os.getenv("USESEND_BASE_URL") or os.getenv("UNSEND_BASE_URL") or DEFAULT_BASE_URL
        if url:
            base = url
        self.url = f"{base}/api/v1"

        self.headers = {
            "Authorization": f"Bearer {self.key}",
            "Content-Type": "application/json",
        }

        self.raise_on_error = raise_on_error
        self._session = session or requests.Session()

        # Lazily initialise resource clients.
        self.emails = Emails(self)
        self.contacts = Contacts(self)
        self.domains = Domains(self)
        self.campaigns = Campaigns(self)

    # ------------------------------------------------------------------
    # Webhooks
    # ------------------------------------------------------------------
    def webhooks(self, secret: str) -> "Webhooks":
        """Create a Webhooks instance for verifying webhook signatures.

        Parameters
        ----------
        secret:
            The webhook signing secret (starts with 'whsec_').

        Returns
        -------
        Webhooks
            A Webhooks instance for verifying signatures and constructing events.

        Example
        -------
        ```python
        usesend = UseSend("us_12345")
        webhooks = usesend.webhooks("whsec_xxx")

        # In your webhook handler
        event = webhooks.construct_event(body, headers=request.headers)
        ```
        """
        return Webhooks(secret)

    # ------------------------------------------------------------------
    # Internal request helper
    # ------------------------------------------------------------------
    def _build_headers(self, extra: Optional[Dict[str, str]] = None) -> Dict[str, str]:
        headers = dict(self.headers)
        if extra:
            headers.update({k: v for k, v in extra.items() if v is not None})
        return headers

    def _request(
        self,
        method: str,
        path: str,
        json: Optional[Any] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Tuple[Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
        """Perform an HTTP request and return ``(data, error)``."""
        resp = self._session.request(
            method,
            f"{self.url}{path}",
            headers=self._build_headers(headers),
            json=json,
        )
        default_error = {"code": "INTERNAL_SERVER_ERROR", "message": resp.reason}

        if not resp.ok:
            try:
                payload = resp.json()
                error = payload.get("error", default_error)
            except Exception:
                error = default_error
            if self.raise_on_error:
                raise UseSendHTTPError(resp.status_code, error, method, path)
            return None, error

        try:
            return resp.json(), None
        except Exception:
            return None, default_error

    # ------------------------------------------------------------------
    # HTTP verb helpers
    # ------------------------------------------------------------------
    def post(
        self,
        path: str,
        body: Any,
        headers: Optional[Dict[str, str]] = None,
    ) -> Tuple[Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
        return self._request("POST", path, json=body, headers=headers)

    def get(
        self, path: str, headers: Optional[Dict[str, str]] = None
    ) -> Tuple[Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
        return self._request("GET", path, headers=headers)

    def put(
        self,
        path: str,
        body: Any,
        headers: Optional[Dict[str, str]] = None,
    ) -> Tuple[Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
        return self._request("PUT", path, json=body, headers=headers)

    def patch(
        self,
        path: str,
        body: Any,
        headers: Optional[Dict[str, str]] = None,
    ) -> Tuple[Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
        return self._request("PATCH", path, json=body, headers=headers)

    def delete(
        self,
        path: str,
        body: Optional[Any] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Tuple[Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
        return self._request("DELETE", path, json=body, headers=headers)


# Import here to avoid circular dependency during type checking
from .emails import Emails  # noqa: E402  pylint: disable=wrong-import-position
from .contacts import Contacts  # noqa: E402  pylint: disable=wrong-import-position
from .domains import Domains  # type: ignore  # noqa: E402
from .campaigns import Campaigns  # type: ignore  # noqa: E402
from .webhooks import Webhooks  # noqa: E402  pylint: disable=wrong-import-position

File: packages/sdk/src/usesend.ts
import { ErrorResponse } from "../types";
import { Contacts } from "./contact";
import { Emails } from "./email";
import { Domains } from "./domain";
import { Campaigns } from "./campaign";
import { Webhooks } from "./webhooks";

const defaultBaseUrl = "https://app.usesend.com";
// eslint-disable-next-line turbo/no-undeclared-env-vars
const baseUrl = `${process?.env?.USESEND_BASE_URL ?? process?.env?.UNSEND_BASE_URL ?? defaultBaseUrl}/api/v1`;

function isUseSendErrorResponse(error: { error: ErrorResponse }) {
  return error.error.code !== undefined;
}

type RequestOptions = {
  headers?: HeadersInit;
};

export class UseSend {
  private readonly baseHeaders: Headers;

  readonly emails = new Emails(this);
  readonly domains = new Domains(this);
  readonly contacts = new Contacts(this);
  readonly campaigns = new Campaigns(this);
  url = baseUrl;

  constructor(
    readonly key?: string,
    url?: string,
  ) {
    if (!key) {
      if (typeof process !== "undefined" && process.env) {
        this.key = process.env.USESEND_API_KEY ?? process.env.UNSEND_API_KEY;
      }

      if (!this.key) {
        throw new Error(
          'Missing API key. Pass it to the constructor `new UseSend("us_123")`',
        );
      }
    }

    if (url) {
      this.url = `${url}/api/v1`;
    }

    this.baseHeaders = new Headers({
      Authorization: `Bearer ${this.key}`,
      "Content-Type": "application/json",
    });
  }

  private mergeHeaders(extra?: HeadersInit) {
    const headers = new Headers(this.baseHeaders);
    if (!extra) {
      return headers;
    }

    const additional = new Headers(extra);
    additional.forEach((value, key) => {
      headers.set(key, value);
    });

    return headers;
  }

  async fetchRequest<T>(
    path: string,
    options: RequestInit = {},
  ): Promise<{ data: T | null; error: ErrorResponse | null }> {
    const requestOptions: RequestInit = {
      ...options,
      headers: this.mergeHeaders(options.headers),
    };

    const response = await fetch(`${this.url}${path}`, requestOptions);
    const defaultError = {
      code: "INTERNAL_SERVER_ERROR",
      message: response.statusText,
    };

    if (!response.ok) {
      try {
        const resp = await response.json();
        if (isUseSendErrorResponse(resp)) {
          return { data: null, error: resp };
        }

        return { data: null, error: resp.error };
      } catch (err) {
        if (err instanceof Error) {
          return {
            data: null,
            error: defaultError,
          };
        }

        return { data: null, error: defaultError };
      }
    }

    const data = await response.json();
    return { data, error: null };
  }

  async post<T>(path: string, body: unknown, options?: RequestOptions) {
    const requestOptions: RequestInit = {
      method: "POST",
      body: JSON.stringify(body),
    };

    if (options?.headers) {
      requestOptions.headers = options.headers;
    }

    return this.fetchRequest<T>(path, requestOptions);
  }

  async get<T>(path: string, options?: RequestOptions) {
    const requestOptions: RequestInit = {
      method: "GET",
    };

    if (options?.headers) {
      requestOptions.headers = options.headers;
    }

    return this.fetchRequest<T>(path, requestOptions);
  }

  async put<T>(path: string, body: any, options?: RequestOptions) {
    const requestOptions: RequestInit = {
      method: "PUT",
      body: JSON.stringify(body),
    };

    if (options?.headers) {
      requestOptions.headers = options.headers;
    }

    return this.fetchRequest<T>(path, requestOptions);
  }

  async patch<T>(path: string, body: any, options?: RequestOptions) {
    const requestOptions: RequestInit = {
      method: "PATCH",
      body: JSON.stringify(body),
    };

    if (options?.headers) {
      requestOptions.headers = options.headers;
    }

    return this.fetchRequest<T>(path, requestOptions);
  }

  async delete<T>(path: string, body?: unknown, options?: RequestOptions) {
    const requestOptions: RequestInit = {
      method: "DELETE",
    };

    if (body !== undefined) {
      requestOptions.body = JSON.stringify(body);
    }

    if (options?.headers) {
      requestOptions.headers = options.headers;
    }

    return this.fetchRequest<T>(path, requestOptions);
  }

  /**
   * Creates a webhook handler with the given secret.
   * Follows the Stripe pattern: `usesend.webhooks(secret).constructEvent(...)`
   *
   * @param secret - Webhook signing secret from your UseSend dashboard
   * @returns Webhooks instance for verifying webhook events
   *
   * @example
   * ```ts
   * const usesend = new UseSend('us_xxx');
   * const webhooks = usesend.webhooks('whsec_xxx');
   *
   * // In your webhook route
   * const event = webhooks.constructEvent(req.body, {
   *   headers: req.headers
   * });
   * ```
   */
  webhooks(secret: string): Webhooks {
    return new Webhooks(secret);
  }
}
